core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -713,6 +760,9 @@ class ReplicaManager(val config: KafkaConfig,
             case HostedPartition.Offline =>
               throw new KafkaStorageException(s"Partition $topicPartition is offline")
 
+            case HostedPartition.Deferred(_) =>
+              throw new KafkaStorageException(s"Partition $topicPartition is deferred")
```

Comment:
Hmm, this doesn't seem right. Not sure why a deferred change would cause a storage error. How about we throw `IllegalStateException` instead to make it clear that this is an unexpected state for now.